include/npcomp/Dialect/ATen/ATenDialect.h

```diff
@@ -0,0 +1,126 @@
+//===- ATenDialect.h --------------------------------------------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef NPCOMP_DIALECT_ATEN_DIALECT_H
+#define NPCOMP_DIALECT_ATEN_DIALECT_H
+
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/Dialect.h"
+#include "mlir/IR/Function.h"
+#include "mlir/IR/OpDefinition.h"
+#include "mlir/IR/OpImplementation.h"
+#include "mlir/IR/StandardTypes.h"
+#include "mlir/IR/TypeSupport.h"
+#include "mlir/IR/Types.h"
+#include "mlir/Interfaces/SideEffectInterfaces.h"
+
+#include <map>
+
+namespace mlir {
+namespace NPCOMP {
+namespace aten {
+
+// The Dialect
+class ATenDialect : public mlir::Dialect {
+public:
+  explicit ATenDialect(mlir::MLIRContext *ctx);
+  static StringRef getDialectNamespace() { return "aten"; }
+
+  /// Parse a type registered to this dialect. Overridding this method is
+  /// required for dialects that have custom types.
+  /// Technically this is only needed to be able to round-trip to textual IR.
+  mlir::Type parseType(DialectAsmParser &parser) const override;
+
+  /// Print a type registered to this dialect. Overridding this method is
+  /// only required for dialects that have custom types.
+  /// Technically this is only needed to be able to round-trip to textual IR.
+  void printType(mlir::Type type, DialectAsmPrinter &os) const override;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+/////////////////////// Custom Types for the Dialect ///////////////////////////
+////////////////////////////////////////////////////////////////////////////////
+
+namespace detail {
+struct ATenListTypeStorage;
+}
+
+/// LLVM-style RTTI: one entry per subclass to allow dyn_cast/isa.
+enum ATenTypeKind {
+  // The enum starts at the range reserved for this dialect.
+  ATEN_TYPE = mlir::Type::FIRST_PRIVATE_EXPERIMENTAL_0_TYPE,
```

Comment:
We have an in-npcomp type range that we should branch off of, but since these are being refactored away upstream, let's not bother updating this.

Reply:
Yes, I didn't consider integrating this high priority.

Reply:
moved to issue